app/src/main/java/com/github/se/travelpouch/ui/travel/ParticipantListScreen.kt

```diff
@@ -214,3 +205,52 @@ fun ParticipantListScreen(
     }
   }
 }
+
+fun handleRoleChange(
+    context: Context,
+    selectedTravel: TravelContainer?,
+    participant: Map.Entry<fsUid, Profile>,
+    newRole: Role,
+    listTravelViewModel: ListTravelViewModel,
+    notificationViewModel: NotificationViewModel,
+    profileViewModel: ProfileModelView,
+    setExpandedRoleDialog: (Boolean) -> Unit,
+    setExpanded: (Boolean) -> Unit
+) {
+  val oldRole = selectedTravel!!.allParticipants[Participant(participant.key)]
+  if (oldRole == newRole) {
+    Toast.makeText(context, "The role is already set to $newRole", Toast.LENGTH_LONG).show()
+    setExpandedRoleDialog(false)
+    setExpanded(false)
+    return
+  }
+  // we have an actual role change
+  if (oldRole == Role.OWNER &&
+      selectedTravel.allParticipants.values.count { it == Role.OWNER } == 1) {
+    Toast.makeText(
+            context,
+            "You're trying to change the role of the only owner of the travel. Please name another owner before changing the role.",
+            Toast.LENGTH_LONG)
+        .show()
+    setExpandedRoleDialog(false)
+    setExpanded(false)
+    return
+  }
+  notificationViewModel.sendNotification(
+      Notification(
+          listTravelViewModel.getNewUid(),
+          profileViewModel.profile.value.fsUid,
+          participant.key,
+          selectedTravel.fsUid,
+          NotificationContent.RoleChangeNotification(selectedTravel.title, newRole),
+          NotificationType.ROLE_UPDATE,
+      ))
+  val participantMap = selectedTravel.allParticipants.toMutableMap()
+  participantMap[Participant(participant.key)] = newRole
+  val updatedContainer = selectedTravel.copy(allParticipants = participantMap.toMap())
+  listTravelViewModel.updateTravel(updatedContainer)
+  listTravelViewModel.selectTravel(updatedContainer)
+  setExpandedRoleDialog(false)
+  setExpanded(false)
+  listTravelViewModel.fetchAllParticipantsInfo()
```

Comment:
I don't understand this part. The notification is sent just before. But here, if I understand,  we are modifiying the role of the user no matter what ?

Reply:
This part of the code send a notification and change the role of a participant. This should be the last case after testing security elements. There are of course security loopholes in this approach, but security would be done in a future PR. Let's say it's a model. This RP is too big to add any more security.

Reply:
Ok, we can correct the security later. However, this needs to be marked as an issue, to avoid forgetting it